components/inputs/input-number.js

```diff
@@ -81,32 +70,69 @@ class InputNumber extends LitElement {
 	}
 
 	render() {
-		const ariaRequired = this.required ? 'true' : undefined;
-
 		return html`
-			<label
-				class="${this.label && !this.labelHidden ? 'd2l-input-label' : 'd2l-offscreen'}"
-				for="${this._inputId}">${this.label}</label>
-			<input
-				aria-required="${ifDefined(ariaRequired)}"
+			<d2l-input-text
 				autocomplete="${ifDefined(this.autocomplete)}"
 				?autofocus="${this.autofocus}"
 				@change="${this._handleChange}"
-				class="d2l-input"
+				@d2l-form-element-should-validate="${this._handleNestedFormElementValidation}"
 				?disabled="${this.disabled}"
+				.forceInvalid="${this.invalid}"
 				id="${this._inputId}"
+				label="${this.label}"
+				?label-hidden="${this.labelHidden}"
 				name="${ifDefined(this.name)}"
 				placeholder="${ifDefined(this.placeholder)}"
-				type="text"
-				.value="${this._formattedValue}">
+				?required="${this.required}"
+				.value="${this._formattedValue}"
+			></d2l-input-text>
+			${ this.validationError ? html`<d2l-tooltip for=${this._inputId} state="error" align="start">${this.validationError}</d2l-tooltip>` : null }
 		`;
 	}
 
+	focus() {
+		this.shadowRoot.querySelector('d2l-input-text').focus();
+	}
+
+	async validate(showErrors) {
+		const errors = await super.validate(showErrors);
+		if (errors.length !== 0) {
+			return errors;
+		}
+		return await this.shadowRoot.querySelector('d2l-input-text').validate(showErrors);
+	}
+
+	get validationMessage() {
+		if (this.validity.rangeOverflow || this.validity.rangeUnderflow) {
+			const label = this.label && !this.labelHidden ? this.label : 'Number';
+			const minNumber = (this.min || this.min === 0) ? formatValue(this.min, this.minFractionDigits, this.maxFractionDigits) : null;
+			const maxNumber = (this.max || this.max === 0) ? formatValue(this.max, this.minFractionDigits, this.maxFractionDigits) : null;
+			if (minNumber && maxNumber) {
+				return this.localize('components.input-number.errorOutsideRange', { label, minNumber, maxNumber });
+			} else if (maxNumber) {
+				return this.localize('components.input-number.errorMaxNumberOnly', { label, maxNumber });
+			} else if (minNumber) {
+				return this.localize('components.input-number.errorMinNumberOnly', { label, minNumber });
+			}
+		}
+		return super.validationMessage;
+	}
+
 	async _handleChange(e) {
 		const value = e.target.value;
 		this._formattedValue = value;
 		await this.updateComplete;
 		this.value = parseNumber(value);
+		this.setFormValue(this.value);
+		this.setValidity({
+			rangeUnderflow: (this.min || this.min === 0) && this.value < this.min,
+			rangeOverflow: (this.max || this.max === 0) && this.value > this.max
```

Comment:
Could maybe do `typeof(this.min) === 'number' && this.value < this.min` to remove some ambiguity around casting `this.min` to a boolean resulting in `false` if it's `0`.